Smt/Query.lean

```diff
@@ -51,7 +51,8 @@ def processVertex (e : Expr) : StateT Solver MetaM Unit := do
     return
   let t ← inferType e
   trace[Smt.debug.query] "t: {t}"
-  let s ← Transformer.exprToTerm t
+  let (s, defs) ← Transformer.exprToTerm t
+  solver := ⟨ solver.commands ++ defs ⟩
```

Comment:
I wonder if we could use this function and the one above to define the `natSub` instead of doing so while processing terms:
- `buildDependencyGraph` takes a set of propositions (including the current goal) as input and builds a dependency graph of symbols (free variables and constants) that _must_ be defined in the SMT query first before asserting those propositions.
- `processVertex` for the lack of a better name (I am open for suggestions) does the actual job of adding each vertex in the dependency graph to the current SMT query. So this is the function in charge of introducing the symbols.
- `generateQuery` calls `buildDependencyGraph` to generate the graph and calls `processVertex` to generate the SMT query. It calls `processVertex` in a depth-first post-fix order to ensure all symbols are introduced in the correct order.

If we remove `natSub` from the set of known constants, it will be added to the dependency graph. Then, we can manually process it in `processVertex`. There are two benefits to this approach:
1. We can avoid duplicate definitions of `natSub` if it's used twice in the goal or once in the goal and once in a hypothesis).
2. If we add `Nat` as a dependency of `natSub`, then we change `natSub`'s inputs from `Int` to `Nat` without having to worry about `Nat` not being defined first.